db/version_set.cc

```diff
@@ -5721,7 +5755,177 @@ Status VersionSet::TryRecoverFromOneManifest(
+      uint32_t cf_id = pair.first;
+      const auto* vstorage =
+          column_family_set_->GetColumnFamily(cf_id)->current()->storage_info();
+      assert(vstorage);
+
+      max_epoch_number =
+          std::max(max_epoch_number, vstorage->GetMaxEpochNumberOfFiles());
+    }
+  }
+
+  SetCurrentNextEpochNumber(max_epoch_number + 1);
+}
+
+Status VersionSet::InferMissingEpochNumber(
+    const std::unordered_map<uint32_t, bool>& cf_to_has_missing_epoch_number) {
+  Status s;
+
+  for (auto& pair : cf_to_has_missing_epoch_number) {
+    bool has_missing_epoch_number = pair.second;
+    if (has_missing_epoch_number) {
+      uint32_t cf_id = pair.first;
+      const auto* vstorage =
+          column_family_set_->GetColumnFamily(cf_id)->current()->storage_info();
+      assert(vstorage);
+
+      std::vector<FileMetaData*> files;
+      for (int level = 0; level < vstorage->num_levels(); ++level) {
+        for (FileMetaData* f : vstorage->LevelFiles(level)) {
+          files.push_back(f);
+        }
+      }
+      InferEpochNumbersFromSeqNo(files);
+      ROCKS_LOG_WARN(db_options()->info_log,
+                     "CF(%d)'s epoch numbers are inferred based on seqno",
+                     cf_id);
+      s = InstallInferredEpochNumberResult(cf_id);
+      ROCKS_LOG_WARN(db_options()->info_log,
+                     "Inferred CF(%d)'s epoch numbers installed in the latest "
+                     "version of the CF",
+                     cf_id);
+      if (!s.ok()) {
+        return s;
+      }
+    }
+  }
+  return s;
+}
+
+Status VersionSet::InstallInferredEpochNumberResult(uint32_t cf_id) {
+  Status s;
+
+  ColumnFamilyData* cfd = column_family_set_->GetColumnFamily(cf_id);
+  assert(cfd->initialized());
+  const auto* current = cfd->current();
+  assert(current);
+  const auto* vstorage = current->storage_info();
+  assert(vstorage);
+
+  VersionEdit file_deletions;
+  VersionEdit file_additions;
+  for (int level = 0; level < vstorage->num_levels(); ++level) {
+    for (FileMetaData* file : vstorage->LevelFiles(level)) {
+      file_deletions.DeleteFile(level, file->fd.GetNumber());
+      file_additions.AddFile(
+          level, file->fd.GetNumber(), file->fd.GetPathId(),
+          file->fd.GetFileSize(), file->smallest, file->largest,
+          file->fd.smallest_seqno, file->fd.largest_seqno,
+          file->marked_for_compaction, file->temperature,
+          file->oldest_blob_file_number, file->oldest_ancester_time,
+          file->file_creation_time, file->epoch_number, file->file_checksum,
+          file->file_checksum_func_name, file->unique_id);
+    }
+  }
+
+  std::unique_ptr<BaseReferencedVersionBuilder> del_builder_guard(
+      new BaseReferencedVersionBuilder(cfd));
+  VersionBuilder* del_builder = del_builder_guard->version_builder();
+  s = del_builder->Apply(&file_deletions);
+  if (s.ok()) {
+    auto* v = new Version(
+        cfd, this, file_options_, *cfd->GetLatestMutableCFOptions(), io_tracer_,
```

Comment:
[TODO] Use EpochNumberRequirement::kMightMissing here and eliminate InTransition enum

Reply:
Fixed